### ISSUE TITLE
Copy of entire logits incur large memory usage at first call of generate. (from v4.45.0)

### ISSUE THREAD

**Issue by kjohew** [huggingface/transformers#34796]
### System Info

- `transformers` version: 4.46.2
- Platform: Linux-5.15.0-122-generic-x86_64-with-glibc2.35
- Python version: 3.10.15
- Huggingface_hub version: 0.26.2
- Safetensors version: 0.4.5
- PyTorch version (GPU?): 2.5.1+cu124 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes

### Who can help?

@zucchini-nlp

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The problem is included since [this commit](https://github.com/huggingface/transformers/blob/19d58d31f19049e8280ccb62a5b098d89909bf5a/src/transformers/generation/utils.py#L3015). Before that, the logits were sliced first and then copied, which would only incur negligible memory overhead. But now, the entire logits will be copied and lead to double large memory usage during the first call of `generate()`.

```
# Previous
next_token_logits = outputs.logits[:, -1, :].clone().float()

# Current
next_token_logits = outputs.logits.clone()[:, -1, :].float()
```

### Expected behavior

hope this problem can be fixed. 🤗

**Comment by zucchini-nlp** [comment]
@kjohew right, seems like I made some changes and reverted them incorrectly. Would you like to open a PR to fix it?

**Comment by kjohew** [comment]
Hi @zucchini-nlp, thanks for response. sure, I will submit a PR later today.

**Comment by zucchini-nlp** [comment]
Lets close it as the PR was merged :)
